code/modules/xenoarchaeology/xenoartifact.dm

```diff
@@ -0,0 +1,489 @@
+    if(!material)
+        material = pick(BLUESPACE, PLASMA, URANIUM, BANANIUM)
+
+    switch(material)
+        if(BLUESPACE)
+            name = "Bluespace [name]"
+            generate_traits(list(/datum/xenoartifact_trait/minor/sharp, /datum/xenoartifact_trait/minor/radioactive,
+                            /datum/xenoartifact_trait/minor/sentient, /datum/xenoartifact_trait/major/sing, 
+                            /datum/xenoartifact_trait/major/laser, /datum/xenoartifact_trait/major/bomb,
+                            /datum/xenoartifact_trait/major/handmore, /datum/xenoartifact_trait/major/emp))
+            if(!xenop.price)
+                xenop.price = pick(100, 200, 300)
+
+        if(PLASMA)
+            name = "Plasma [name]"
+            generate_traits(list(/datum/xenoartifact_trait/major/sing, /datum/xenoartifact_trait/activator/burn,
+                            /datum/xenoartifact_trait/minor/dense, /datum/xenoartifact_trait/minor/sentient, 
+                            /datum/xenoartifact_trait/major/capture, /datum/xenoartifact_trait/major/timestop,
+                            /datum/xenoartifact_trait/major/bomb, /datum/xenoartifact_trait/major/mirrored,
+                            /datum/xenoartifact_trait/major/corginator,/datum/xenoartifact_trait/activator/clock,
+                            /datum/xenoartifact_trait/major/invisible,/datum/xenoartifact_trait/major/handmore,
+                            /datum/xenoartifact_trait/major/lamp, /datum/xenoartifact_trait/major/forcefield,
+                            /datum/xenoartifact_trait/activator/signal,/datum/xenoartifact_trait/major/heal))
+            if(!xenop.price)
+                xenop.price = pick(200, 300, 500)
+            malfunction_mod = 2
+
+        if(URANIUM)
+            name = "Uranium [name]"
+            generate_traits(list(/datum/xenoartifact_trait/major/sing, /datum/xenoartifact_trait/minor/sharp,
+                            /datum/xenoartifact_trait/major/laser, /datum/xenoartifact_trait/major/corginator,
+                            /datum/xenoartifact_trait/minor/sentient, /datum/xenoartifact_trait/minor/wearable,
+                            /datum/xenoartifact_trait/major/handmore, /datum/xenoartifact_trait/major/invisible,
+                            /datum/xenoartifact_trait/major/heal), TRUE) 
+            if(!xenop.price)
+                xenop.price = pick(300, 500, 800) 
+            malfunction_mod = 8
+
+        if(BANANIUM)
+            name = "Australium [name]"
+            generate_traits(list(/datum/xenoartifact_trait/major/sing))
+            if(!xenop.price)
+                xenop.price = pick(500, 800, 1000) 
+            malfunction_mod = 0.5
+
+    icon_state = null
+    for(var/datum/xenoartifact_trait/T in traits) //This is kinda weird but it stops certain runtime cases.
+        if(istype(T, /datum/xenoartifact_trait/minor/dense))
+            T.on_init(src)
+            return
```

Comment:
Not at all! This stops other traits generating before the dense trait, and causing major issues. The dense trait makes the artifact into a structure, which handles its own icon, since it uses a different set of sprites, bigger ones.